tests/distributions/test_gaussian_mixtures.py

```diff
@@ -0,0 +1,108 @@
+    n_samples = 2000 * 1000
+    if batch_mode:
+        sample_shape = torch.Size(())
+    else:
+        sample_shape = torch.Size((n_samples,))
+    locs = torch.tensor(torch.rand(K, D), requires_grad=True)
+    sigmas = torch.ones(D) + 0.5 * torch.rand(D)
+    sigmas = torch.tensor(sigmas, requires_grad=True)
+    if not flat_logits:
+        logits = torch.tensor(1.5 * torch.rand(K), requires_grad=True)
+    else:
+        logits = torch.tensor(0.1 * torch.rand(K), requires_grad=True)
+    omega = torch.tensor(0.2 * torch.ones(D) + 0.1 * torch.rand(D), requires_grad=False)
+
+    _pis = torch.exp(logits)
+    pis = _pis / _pis.sum()
+
+    if cost_function == 'cosine':
+        analytic1 = torch.cos((omega * locs).sum(-1))
+        analytic2 = torch.exp(-0.5 * torch.pow(omega * sigmas, 2.0).sum(-1))
+        analytic = (pis * analytic1 * analytic2).sum()
+        analytic.backward()
+    elif cost_function == 'quadratic':
+        analytic = torch.pow(sigmas, 2.0).sum(-1) + torch.pow(locs, 2.0).sum(-1)
+        analytic = (pis * analytic).sum()
+        analytic.backward()
+
+    analytic_grads = {}
+    analytic_grads['locs'] = locs.grad.clone()
+    analytic_grads['scale'] = sigmas.grad.clone()
+    analytic_grads['logits'] = logits.grad.clone()
+    assert locs.grad.shape == locs.shape
+    assert sigmas.grad.shape == sigmas.shape
+    assert logits.grad.shape == logits.shape
+
+    sigmas.grad.zero_()
+    logits.grad.zero_()
+    locs.grad.zero_()
+
+    if mix_dist == MixtureOfDiagNormalsSharedCovariance:
+        params = {'locs': locs, 'scale': sigmas, 'logits': logits}
+        if batch_mode:
+            locs = locs.unsqueeze(0).expand(n_samples, K, D)
+            sigmas = sigmas.unsqueeze(0).expand(n_samples, D)
+            logits = logits.unsqueeze(0).expand(n_samples, K)
+            dist_params = {'locs': locs, 'scale': sigmas, 'logits': logits}
+        else:
+            dist_params = params
+
+    dist = mix_dist(**dist_params)
+    z = dist.rsample(sample_shape=sample_shape)
+    assert z.shape == (n_samples, D)
+    if cost_function == 'cosine':
+        cost = torch.cos((omega * z).sum(-1)).sum() / float(n_samples)
+    elif cost_function == 'quadratic':
+        cost = torch.pow(z, 2.0).sum() / float(n_samples)
+    cost.backward()
+
+    assert_equal(analytic, cost, prec=0.1,
+                 msg='bad cost function evaluation for {} test (expected {}, got {})'.format(
+                     mix_dist.__name__, analytic.item(), cost.item()))
+    print("analytic_grads_logit", analytic_grads['logits'].detach().numpy())
+
+    for param_name, param in params.items():
+        assert_equal(param.grad, analytic_grads[param_name], prec=0.06,
+                     msg='bad {} grad for {} (expected {}, got {})'.format(
+                         param_name, mix_dist.__name__, analytic_grads[param_name], param.grad))
+
+
+@pytest.mark.parametrize('batch_size', [1, 3])
```

Comment:
ditto: strengthen this via something like
```py
@pytest.mark.parametrize('sample_shape', [(), (4,), (3,2)])
```

Reply:
these tests are expensiveish can only run so many variants in CI

Reply:
Ok, I you volunteer to be assigned to any bugs due to broadcasting errors.

Reply:
yes i volunteer